onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -0,0 +1,268 @@
+  auto scratch_data = allocator->Alloc(batch_size * num_heads_ * sequence_length * sequence_length * element_size);
+  BufferUniquePtr scratch_buffer(scratch_data, BufferDeleter(allocator));
+
+  {
+    memset(scratch_data, 0, batch_size * num_heads_ * sequence_length * sequence_length * element_size);
+    auto mask_data = mask_index->template Data<int>();
+    int size_each_batch = num_heads_ * sequence_length;
+    T* p_current_data = reinterpret_cast<T*>(scratch_data);
+    for (int b_i = 0; b_i < batch_size; b_i++) {
+      int mask = mask_data[b_i];
+      for (int n_i = 0; n_i < size_each_batch; n_i++) {
+        for (int m_i = mask; m_i < sequence_length; m_i++) {
+          p_current_data[m_i] = static_cast<T>(-10000.0);
+        }
+        p_current_data += sequence_length;
+      }
+    }
+  }
+
+  {
+    int offset_Q = 0;
+    int offset_Q_increment = sequence_length * head_size;
+    int offset_scratch = 0;
+    int offset_scratch_increment = sequence_length * sequence_length;
+
+    for (int b_i = 0; b_i < batch_size; b_i++) {
+      for (int n_i = 0; n_i < num_heads_; n_i++) {
+        math::Gemm<T>(
+            CblasNoTrans,
+            CblasTrans,
+            sequence_length,
+            sequence_length,
+            head_size,
+            1.0f / sqrt(static_cast<float>(head_size)),
+            Q + offset_Q,
+            K + offset_Q,
+            1.0,
+            reinterpret_cast<T*>(scratch_data) + offset_scratch,
+            tp);
+        offset_Q += offset_Q_increment;
+        offset_scratch += offset_scratch_increment;
+      }
+    }
+  }
+
+  // STEP.4: P(B, N, S, S) = Softmax(scratch)
+  auto p_data = allocator->Alloc(batch_size * num_heads_ * sequence_length * sequence_length * element_size);
+  BufferUniquePtr p_buffer(p_data, BufferDeleter(allocator));
+
+  {
+    int N = batch_size * num_heads_ * sequence_length;
+    int D = sequence_length;
+
+    Eigen::TensorMap<Eigen::Tensor<const T, 2, Eigen::RowMajor, Eigen::DenseIndex>, Eigen::Aligned> X_tensor(
+        reinterpret_cast<T*>(scratch_data), N, D);
+    Eigen::TensorMap<Eigen::Tensor<T, 2, Eigen::RowMajor, Eigen::DenseIndex>, Eigen::Aligned> Y_tensor(
+        reinterpret_cast<T*>(p_data), N, D);
+#ifndef USE_OPENMP
+    if (tp == nullptr)
+#endif
+      ComputeSoftMax(Eigen::DefaultDevice(), X_tensor, Y_tensor, false);
+#ifndef USE_OPENMP
+    else
+      ComputeSoftMax(Eigen::ThreadPoolDevice(&tp->GetHandler(), tp->NumThreads()), X_tensor, Y_tensor, false);
+#endif
+  }
+
+  // STEP.5: out_tmp(B, N, S, H) = P(B, N, S, S) x V(B, N, S, H)
+  auto out_tmp_data = allocator->Alloc(batch_size * num_heads_ * sequence_length * head_size * element_size);
+  BufferUniquePtr out_tmp_buffer(out_tmp_data, BufferDeleter(allocator));
+  {
+    int offset_p = 0;
+    int offset_p_increment = sequence_length * sequence_length;
+    int offset_V = 0;
+    int offset_V_increment = sequence_length * head_size;
+
+    for (int b_i = 0; b_i < batch_size; b_i++) {
```

Comment:
Can we use batch GEMM to improve performance? Like https://software.intel.com/en-us/articles/introducing-batch-gemm-operations

Reply:
I am trying different ways to improve the performance of Attention, which including Gemm and MatMul as well. That will come later (not going to be a part of this PR)